Add 'dropwizard-jdbi3' module

FredDeschenes opened the pull request:
###### Problem:
jDBI3 as been released for a couple of months and with the recent update to Metrics 4.X Dropwizard can now use it without losing any of the features from jDBI2. This should of course not be a forced upgrade and allow users to decide which jDBI version they wish to use in their project. jDBI now also supports most of the custom mappers that were in `dropwizard-jdbi` out of the box or in a separate module, removing some code from Dropwizard.

###### Solution:
Create a `dropwizard-jdbi3` module alongside the existing `dropwizard-jdbi` module.

###### Result:
Added a basic implementation of `dropwizard-jdbi3`, without any of the previous mappers or documentation.

TODO:

- [x] Add documentation (how should this be approached? a new page for the module? have documentation for both modules in the current page?)
- [x] Check if any existing mappers need to be implemented for jDBI3 (jDBI has a module for Guava which seems to cover every mapper Dropwizard had, Joda time is effectively deprecated, we might need the OptionalInt/Long/Double/etc)
- [x] Add unit tests for `JdbiFactory`

isaki-x commented:
For `JdbiFactory`, can you please follow the configure paradigm in `io.dropwizard.jdbi.DBIFactory`? I had implemented that to inject custom configuration in my own subclass of DBIFactory. Thanks.

FredDeschenes commented:
@isaki-x : Will do.

FredDeschenes commented:
I validated that all Dropwizard-specific mappers are now not required as they are either supported out of the box by jDBI or one of its plugins (Joda or Guava).

arteam commented:
I think it's better to leave them out. For users who uses Guava and JodaTime, the modules easy to include to their projects.

joschi commented:
@FredDeschenes I'd vote for including these plugins by default because Dropwizard usually comes with "batteries included" for the libraries it's using itself (i. e. JodaTime and Guava).

arteam commented on dropwizard-jdbi3/src/main/java/io/dropwizard/jdbi3/JdbiFactory.java:
```diff
@@ -0,0 +1,118 @@
+                      String name) {
+
+        // Create the instance
+        final Jdbi jdbi = this.newInstance(dataSource);
```
`this` seems to be redundant, because there's no superclass for `JdbiFactory` which has a method with the same name.

FredDeschenes replied:
Removed 'this'

arteam commented on dropwizard-jdbi3/src/main/java/io/dropwizard/jdbi3/NamePrependingTemplateEngine.java:
```diff
@@ -0,0 +1,46 @@
+        final ExtensionMethod extensionMethod = ctx.getExtensionMethod();
+        final String originalResult = originalEngine.render(template, ctx);
+
+        if ((extensionMethod != null)
```
I think it would be more clear to make a check that the extension method _doesn't_ exist and exit early.
```java
if (extensionMethod == null || extensionMethod.getType() == null || extensionMethod.getMethod() == null) {
            return originalResult;
}
```

FredDeschenes replied:
Turns out `ExtensionMethod` has `requireNonNull` for type and method so I'll remove those unnecessary checks as well.

arteam commented on dropwizard-jdbi3/src/main/java/io/dropwizard/jdbi3/NamePrependingTemplateEngine.java:
```diff
@@ -0,0 +1,46 @@
+ * Jdbi {@link TemplateEngine} that prepends the SQLObject's type and method name in front of every
+ * statement sent to the database.
+ */
+public class NamePrependingTemplateEngine implements TemplateEngine {
```
It would be nice to have a simple unit test for `NamePrependingTemplateEngine` to verify that it actually inserts comments to the query.

FredDeschenes replied:
Test added

arteam commented on dropwizard-jdbi3/src/test/java/io/dropwizard/jdbi3/bundles/JDBIExceptionsBundleTest.java:
```diff
@@ -0,0 +1,28 @@
+
+        new JdbiExceptionsBundle().run(environment);
+
+        verify(jerseyEnvironment, times(1)).register(isA(LoggingSQLExceptionMapper.class));
```
`times(1)` seems to be redundant, because `verify` by default checks that the call has happened only once.

arteam commented on dropwizard-jdbi3/src/test/java/io/dropwizard/jdbi3/JdbiHealthCheckTest.java:
```diff
@@ -0,0 +1,45 @@
+        Handle handle = mock(Handle.class);
+        when(jdbi.open()).thenReturn(handle);
+        when(jdbi.withHandle(Mockito.any())).thenCallRealMethod();
+        Mockito.doAnswer(invocation -> {
```
The `handle.execute` method returns `int` instead of `void`, so we can use the `when().doAnswer` form.

FredDeschenes replied:
It's doing this for the `sleep` in the invocation, I don't think I can change this (there's no `when(handle...).doAnswer`, I'd have to do `.thenReturn` which wouldn't sleep). No sure why it's sleeping anyway so I can probably change this. What do you think?

arteam commented on dropwizard-jdbi3/src/test/java/io/dropwizard/jdbi3/JdbiHealthCheckTest.java:
```diff
@@ -0,0 +1,45 @@
+        when(jdbi.withHandle(Mockito.any())).thenCallRealMethod();
+        Mockito.doAnswer(invocation -> {
+            try {
+                TimeUnit.SECONDS.sleep(10);
```
The `answer` method can throw exceptions, so we can avoid catching the exception.

arteam commented on dropwizard-jdbi3/src/test/java/io/dropwizard/jdbi3/JdbiHealthCheckTest.java:
```diff
@@ -0,0 +1,45 @@
+                validationQuery);
+        HealthCheck.Result result = jdbiHealthCheck.check();
+        executorService.shutdown();
+        assertThat("is unhealthy", false, is(result.isHealthy()));
```
Could you please use an AssertJ assertions for consistency with rest of the codebase?

FredDeschenes replied:
Yeah this was copied from the DBIHealthcheck in the jdbi2 module, I'll convert them.

arteam commented on dropwizard-jdbi3/src/main/java/io/dropwizard/jdbi3/jersey/LoggingJDBIExceptionMapper.java:
```diff
@@ -0,0 +1,35 @@
+ * Iterates through a {@link JdbiException}'s cause if it's a {@link SQLException} otherwise log as normal.
+ */
+@Provider
+public class LoggingJdbiExceptionMapper extends LoggingExceptionMapper<JdbiException> {
```
I guess the file also should be named `LoggingJdbiExceptionMapper` instead `LoggingJDBIExceptionMapper`

FredDeschenes replied:
Yeah the joys of file paths on Windows with Git... I'll rename those properly.

arteam commented:
> I'd vote for including these plugins by default because Dropwizard usually comes with "batteries included" for the libraries it's using itself (i. e. JodaTime and Guava).

On second thought, I think it makes sense. The modules are small and it will make transition from `dropwizard-jdbi2` a whole lot simpler.

FredDeschenes commented:
I added the JodaTime and Guava plugins and also added the jDBI3-bom to Dropwizard's bom, but I'm not super used to Maven, just make sure I haven't messed up the dependencies to everything else ;)

FredDeschenes commented:
Concerning the documentation, I thought about having a new 'JDBI3' page with updated documentation content (although it's going to be pretty similar) and flagging the 'jdbi2' page to tell new projects that they should use the new 'jdbi3' module but existing projects can update at their discretion. Sounds good?

arteam commented on dropwizard-jdbi3/src/test/java/io/dropwizard/jdbi3/JdbiHealthCheckTest.java:
```diff
@@ -0,0 +1,65 @@
+        when(jdbi.open()).thenReturn(handle);
+        when(jdbi.withHandle(Mockito.any())).thenCallRealMethod();
+
+        when(handle.execute(validationQuery)).thenAnswer((Answer<Integer>) invocation -> null);
```
Because the method returns an integer, I think we could just use `thenReturn(0)`

arteam commented on dropwizard-jdbi3/src/test/java/io/dropwizard/jdbi3/JdbiHealthCheckTest.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Test
+    public void testNoTimeoutReturnsHealthy() throws Exception {
+        String validationQuery = "select 1";
```
It seems that we could extract preparing of `executorService`, `validationQuery`, `jdbi` and `handle` to a `setup` method and share it between the tests.

arteam commented on dropwizard-jdbi3/src/test/java/io/dropwizard/jdbi3/JdbiHealthCheckTest.java:
```diff
@@ -0,0 +1,65 @@
+            jdbi,
+            validationQuery);
+        HealthCheck.Result result = jdbiHealthCheck.check();
+        executorService.shutdown();
```
I think it would be nice to shutdown the executor in a `tearDown` method, so even if the test fails, we correctly close resources.

arteam commented on dropwizard-jdbi3/src/test/java/io/dropwizard/jdbi3/JdbiHealthCheckTest.java:
```diff
@@ -0,0 +1,65 @@
+
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        JdbiHealthCheck jdbiHealthCheck = new JdbiHealthCheck(executorService,
+            Duration.milliseconds(5),
```
I would increase timeout from 5ms to 100ms. 5ms is too little to get a result from a new thread. The test may fail with a timeout while the new thread is being scheduled to run.

arteam commented on dropwizard-jdbi3/src/test/java/io/dropwizard/jdbi3/JdbiHealthCheckTest.java:
```diff
@@ -0,0 +1,65 @@
+
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        JdbiHealthCheck jdbiHealthCheck = new JdbiHealthCheck(executorService,
+                Duration.milliseconds(5),
```
I would increase the timeout, because a test with 5ms timeout can fail sporadically, not because we actually wait for a result from the database.

arteam commented on dropwizard-jdbi3/src/test/java/io/dropwizard/jdbi3/NamePrependingTemplateEngineTest.java:
```diff
@@ -0,0 +1,60 @@
+        final TemplateEngine original = mock(TemplateEngine.class);
+        final StatementContext ctx = mock(StatementContext.class);
+        final String template = UUID.randomUUID().toString();
+        final String originalRendered = UUID.randomUUID().toString();
```
I think we could extract preparing of the variables to a `setup` method and share it between the tests.

arteam commented on dropwizard-jdbi3/src/main/java/io/dropwizard/jdbi3/JdbiHealthCheck.java:
```diff
@@ -0,0 +1,35 @@
+    @Override
+    protected Result check() throws Exception {
+        return timeBoundHealthCheck.check(() ->
+            jdbi.withHandle((handle) -> {
```
On second thought, I think it would be better to use ` try (Handle handle = jdbi.open())` instead of `withHandle`. The biggest advantages of lambda-expressions is that we could use a short-form for one-line calls or even reference a method. Here we use two commands, so basically the both expressions require the same amount of code. At the same time the former version is easier to test.

arteam commented:
> Concerning the documentation, I thought about having a new 'JDBI3' page with updated documentation content (although it's going to be pretty similar) and flagging the 'jdbi2' page to tell new projects that they should use the new 'jdbi3' module but existing projects can update at their discretion. Sounds good?

Sounds good to me!

arteam commented:
I think we also need to add an integration test test to verify that Dropwizard and Jdbi3 together and you can write database access code with `JdbiFactory`. For example, for Jdbi2 we've got `JDBITest`. In my personal project, I wrote a couple tests which showcase Jdbi features, you can get an idea:
* https://github.com/arteam/dropwizard-jdbi3/blob/master/src/test/java/com/github/arteam/jdbi3/JdbiTest.java
* https://github.com/arteam/dropwizard-jdbi3/blob/master/src/test/java/com/github/arteam/jdbi3/GameDao.java
* https://github.com/arteam/dropwizard-jdbi3/blob/master/src/test/resources/com/github/arteam/jdbi3/GameDao.sql.stg

isaki-x commented:
One note on the configure implementation, you can drop the `PooledDataSourceFactory` argument; the was something I hadn't wanted there in the first place but had no other simple way of getting some of the things required to configure the legacy mappers/translators. There is no harm in leaving it, it just was something added as a shortcut and due to design requirements. Thanks!

arteam commented:
@FredDeschenes Thank you very much for the contribution, great work! I'm going to make a RC shortly.

isaki-x commented:
This is super exciting, I look forward to migrating when this is ready. Thanks for the work @FredDeschenes !
